File: Tyche.iOS/Core/Sources/Core/Paging/CursorPage.swift
public struct CursorPage<Value: Codable>: Codable {
    public let items: [Value]
    public let next: String?
    
    public init(items: [Value], next: String?) {
        self.items = items
        self.next = next
    }
}

public extension CursorPage {
    @inlinable
    func map<OutValue>(_ transform: (Value) -> OutValue) -> CursorPage<OutValue> {
        return CursorPage<OutValue>(
            items: self.items.map(transform),
            next: self.next
        )
    }
}

File: Tyche.Android/core/src/main/java/com/felipearpa/tyche/core/paging/CursorPage.kt
package com.felipearpa.tyche.core.paging

import kotlinx.serialization.Serializable

@Serializable
data class CursorPage<Value : Any>(
    val items: List<Value>,
    val next: String?,
)

inline fun <Value : Any, Target : Any> CursorPage<Value>.map(transform: (Value) -> Target): CursorPage<Target> =
    CursorPage(
        items = this.items.map(transform),
        next = this.next,
    )

private val emptyCursorPageInstance = CursorPage<Nothing>(
    items = emptyList(),
    next = null,
)

fun <Value : Any> emptyCursorPage() = emptyCursorPageInstance
